### DISCUSSION TITLE
Listeners or Subscribers?

### DISCUSSION THREAD

**Discussion by karaka200** [symfony/symfony#45342] [Q&A]
Which is more suitable for use in events? Listeners or Subscribers ?
Please tell me your opinion

**Comment by TimoBakx** [comment]
Both are suitable and each has their pros and cons.

Personally, I prefer invokable listeners (using a single `__invoke()` method), to enforce a single responsibility. The drawback of that is that the configuration of which event (and with which priority) is listened to, is outside the class in the services configuration.

For subscribers, that configuration is inside the class itself, and it's nice to have both those together.
The drawback of subscribers, as from my experience, is that some devs tend to just add more and more methods into a single subscriber, ending up with something like a `UserSubscriber` with a lot of logic in 1 big class.

**Reply by norkunas** [reply]
> The drawback of that is that the configuration of which event (and with which priority) is listened to, is outside the class in the services configuration.

With an `AsEventListener` attribute that can be in a same class :)

**Reply by TimoBakx** [reply]
Interesting. I currently defined a bunch of interfaces and bind those to specific events. Implementing the interface in my listener autoconfigures it to the right event and priority. :)
